x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/CreateEnrollmentToken.java

```diff
@@ -0,0 +1,284 @@
+            throw new IllegalStateException("Error generating enrollment token: " + e.getMessage());
+        }
+    }
+
+    private HttpResponse.HttpResponseBuilder responseBuilder(InputStream is) throws IOException {
+        final HttpResponse.HttpResponseBuilder httpResponseBuilder = new HttpResponse.HttpResponseBuilder();
+        if (is != null) {
+            byte[] bytes = toByteArray(is);
+            String responseBody = new String(bytes, StandardCharsets.UTF_8);
+            logger.debug(responseBody);
+            httpResponseBuilder.withResponseBody(responseBody);
+        } else {
+            logger.debug("<Empty response>");
+        }
+        return httpResponseBuilder;
+    }
+
+    private byte[] toByteArray(InputStream is) throws IOException {
+        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        byte[] internalBuffer = new byte[1024];
+        int read = is.read(internalBuffer);
+        while (read != -1) {
+            baos.write(internalBuffer, 0, read);
+            read = is.read(internalBuffer);
+        }
+        return baos.toByteArray();
+    }
+
+    protected static URL createAPIKeyURL(URL url) throws MalformedURLException, URISyntaxException {
+        return new URL(url, (url.toURI().getPath() + "/_security/api_key").replaceAll("/+", "/"));
+    }
+
+    protected static URL getHttpInfoURL(URL url) throws MalformedURLException, URISyntaxException {
+        return new URL(url, (url.toURI().getPath() + "/_nodes/http").replaceAll("/+", "/"));
+    }
+
+    protected static List<String> getBoundAddresses(Map<?, ?> nodesInfo) {
+        nodesInfo = (Map<?, ?>) nodesInfo.get("nodes");
+        Map<?, ?> nodeInfo = (Map<?, ?>) nodesInfo.values().iterator().next();
+        Map<?, ?> http = (Map<?, ?>) nodeInfo.get("http");
+        return (ArrayList<String>) http.get("bound_address");
+    }
+
+    static String getVersion(Map<?, ?> nodesInfo) {
+        nodesInfo = (Map<?, ?>) nodesInfo.get("nodes");
+        Map<?, ?> nodeInfo = (Map<?, ?>) nodesInfo.values().iterator().next();
+        return nodeInfo.get("version").toString();
+    }
+
+    protected String getApiKey(String user, SecureString password) throws Exception {
+        final CheckedSupplier<String, Exception> createApiKeyRequestBodySupplier = () -> {
+            XContentBuilder xContentBuilder = JsonXContent.contentBuilder();
+            xContentBuilder.startObject()
+                .field("name", "enrollment_token_API_key_" + UUIDs.base64UUID())
+                .field("expiration", ENROLL_API_KEY_EXPIRATION)
+                .startObject("role_descriptors")
+                .startObject("create_enrollment_token")
+                .field("cluster", "[cluster:admin/xpack/security/enrollment/enroll/node]")
+                .endObject()
+                .endObject()
+                .endObject();
+            return Strings.toString(xContentBuilder);
+        };
+
+        final URL url = new URL(client.getDefaultURL());
```

Comment:
I'm wondering wether we should be calling `getDefaultURL()` here every time given the process it takes to determine the URL. Part of it is redundant ( we know it's gonna be https for instance ) but mostly wondering should we just call it once for this class ?